Critical compiler error related to serde/recent alias analysis changes

I was testing Enso with network throttling and encountered a critical blocker:
![Screenshot from 2024-09-06 16-20-40](https://github.com/user-attachments/assets/08d8ffa1-4ea4-4548-b527-131041855a75)

Started in the logs with:
```
[WARN] [2024-09-06T16:14:19.147] [enso.org.enso.interpreter.service.ExecutionService] Error while ensureCompiledModule local.New_Project_6.Main
java.io.IOException: Adding at 12211471 object:
  org.enso.compiler.data.BindingsMap$ResolvedImport@6da4c80a
but there already is:
  org.enso.compiler.data.BindingsMap$ResolvedImport@65b1ac18
are they equal: false
	at org.enso.runtime/org.enso.persist.PerInputImpl.readIndirect(PerInputImpl.java:229)
	at org.enso.runtime/org.enso.persist.PerInputImpl.readObject(PerInputImpl.java:74)
	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:227)
	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:207)
	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
	at org.enso.runtime/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
	at org.enso.runtime/org.enso.interpreter.caches.ImportExportCache$PersistBindingsMap.readObject(ImportExportCache.java:237)
	at org.enso.runtime/org.enso.interpreter.caches.ImportExportCache$PersistBindingsMap.readObject(ImportExportCache.java:197)
	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
	at org.enso.runtime/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
	at org.enso.runtime/org.enso.persist.PerBufferReference.readObject(PerBufferReference.java:48)
	at org.enso.runtime/org.enso.persist.Persistance$Reference.get(Persistance.java:253)
	at org.enso.runtime/org.enso.compiler.core.ir.IrLazyMap$En.getValue(IrLazyMap.java:57)
	at org.enso.runtime/org.enso.compiler.core.ir.IrLazyMap.get(IrLazyMap.java:37)
	at org.enso.runtime/org.enso.compiler.core.ir.MetadataStorage.get(MetadataStorage.java:81)
	at org.enso.runtime/org.enso.compiler.core.Implicits$AsMetadata.unsafeGetMetadata(Implicits.scala:108)
	at org.enso.runtime/org.enso.compiler.data.BindingsMap$ResolvedModule.exportedSymbols(BindingsMap.scala:760)
```

only to get worse with
```
[ERROR] [2024-09-06T16:14:19.157] [enso.org.enso.interpreter.service.ExecutionService] Error executing org.enso.interpreter.instrument.job.EnsureCompiledJob@4dcc8141
org.enso.compiler.core.CompilerError: Compiler Internal Error: Empty dataflow analysis metadata during changeset calculation.
	at org.enso.runtime/org.enso.compiler.core.Implicits$AsMetadata.$anonfun$unsafeGetMetadata$1(Implicits.scala:109)
	at org.enso.runtime/scala.Option.getOrElse(Option.scala:201)
	at org.enso.runtime/org.enso.compiler.core.Implicits$AsMetadata.unsafeGetMetadata(Implicits.scala:109)
	at org.enso.runtime/org.enso.interpreter.instrument.ChangesetBuilder.compute(ChangesetBuilder.scala:128)
	at org.enso.runtime/org.enso.interpreter.instrument.ChangesetBuilder.build(ChangesetBuilder.scala:113)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.$anonfun$applyEdits$2(EnsureCompiledJob.scala:363)
```

and worse
```
[ERROR] [2024-09-06T16:14:19.222] [enso.org.enso.interpreter.runtime.Module] Cannot read BindingsMap for Standard.Base.Main: Adding at 12199680 object:
  org.enso.compiler.data.BindingsMap$ResolvedImport@5f445064
but there already is:
  org.enso.compiler.data.BindingsMap$ResolvedImport@1724f030
are they equal: false
[DEBUG] [2024-09-06T16:14:19.222] [enso.org.enso.interpreter.runtime.Module] Cannot read BindingsMap for Standard.Base.Main: Adding at 12199680 object:
  org.enso.compiler.data.BindingsMap$ResolvedImport@5f445064
but there already is:
  org.enso.compiler.data.BindingsMap$ResolvedImport@1724f030
are they equal: false
java.io.IOException: Adding at 12199680 object:
  org.enso.compiler.data.BindingsMap$ResolvedImport@5f445064
but there already is:
  org.enso.compiler.data.BindingsMap$ResolvedImport@1724f030
are they equal: false
	at org.enso.runtime/org.enso.persist.PerInputImpl.readIndirect(PerInputImpl.java:229)
	at org.enso.runtime/org.enso.persist.PerInputImpl.readObject(PerInputImpl.java:74)
	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:227)
	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:207)
```

and eventually start reporting
```
[ERROR] [2024-09-06T16:14:20.455] [enso.org.enso.interpreter.service.ExecutionService] Evaluation of visualization [Text(Standard.Visualization.Helpers,a -> a.default_visualization.to_js_object.to_json,Vector())] failed in module [Module[Standard.Visualization.Helpers]] with [class java.lang.IllegalStateException]: Unexpected FrameAnalysisMeta associated with IR DefinitionArgument.Specified(name = Name.Self(location = None, synthetic = true, passData = MetadataStorage[], diagnostics = null, id = null), ascribedType = None, defaultValue = None, suspended = false, location = None, passData = MetadataStorage[AliasMetadata.Occurrence, DataflowAnalysis.DependencyInfo, FramePointer, IgnoredBindings.State.Ignored, TailCall.TailPosition.NotTail], diagnostics = null, id = 581d1b35-2c9f-42ff-92e2-0deb5d55cbc9)
Old: FramePointer[parentLevel=0, frameSlotIdx=1] new FramePointer[parentLevel=0, frameSlotIdx=1]
[ERROR] [2024-09-06T16:14:20.459] [enso.org.enso.interpreter.service.ExecutionService] Visualization for expression 6d61c10e-d709-4e13-a7e8-387f2d714e10 failed: Unexpected FrameAnalysisMeta associated with IR DefinitionArgument.Specified(name = Name.Self(location = None, synthetic = true, passData = MetadataStorage[], diagnostics = null, id = null), ascribedType = None, defaultValue = None, suspended = false, location = None, passData = MetadataStorage[AliasMetadata.Occurrence, DataflowAnalysis.DependencyInfo, FramePointer, IgnoredBindings.State.Ignored, TailCall.TailPosition.NotTail], diagnostics = null, id = 581d1b35-2c9f-42ff-92e2-0deb5d55cbc9)
Old: FramePointer[parentLevel=0, frameSlotIdx=1] new FramePointer[parentLevel=0, frameSlotIdx=1] (evaluation result: None)
```

Note sure if the network throttling is important (in chrome, Slow 4G) but I haven't seen it otherwise.
Full logs: 
[out2.txt](https://github.com/user-attachments/files/16909358/out2.txt)
Backend built with fa1e14945a9d6933abfecc8993c8f1c472992537

--- a/lib/java/persistance/src/main/java/org/enso/persist/PerInputImpl.java
+++ b/lib/java/persistance/src/main/java/org/enso/persist/PerInputImpl.java
@@ -217,22 +217,7 @@ final class PerInputImpl implements Input {
     var inData = new PerInputImpl(cache, at);
     var res = p.readWith(inData);
     res = cache.resolveObject(res);
-    var prev = cache.putObjectAt(at, res);
-    if (prev != null) {
-      var bothObjectsAreTheSame = Objects.equals(res, prev);
-      var sb = new StringBuilder();
-      sb.append("Adding at ").append(at).append(" object:\n  ");
-      dumpObject(sb, res);
-      sb.append("\nbut there already is:\n  ");
-      dumpObject(sb, prev);
-      sb.append("\nare they equal: ").append(bothObjectsAreTheSame);
-      var ex = new IOException(sb.toString());
-      if (bothObjectsAreTheSame) {
-        PerUtils.LOG.warn(sb.toString(), ex);
-      } else {
-        throw raise(RuntimeException.class, ex);
-      }
-    }
+    cache.putObjectAt(at, res);
     return res;
   }
 

